pubsub/gossipsub/gossipsub-v1.2.md

```diff
@@ -29,32 +49,95 @@ This section lists the configuration parameters that control the behaviour of th
 | Parameter | Description  | Reasonable Default |
 | -------- | -------- | -------- |
 | `D_non_choke`     | The minimum number of peers in a mesh that must remain unchoked. | `D_lo` |
-| `choke_heartbeat_interval` | The number of heartbeats before assessing and applying `CHOKE`/`UNCHOKE` control messages and adding `D_max_add`. | 20 |
-| `D_max_add` | The maximum number of peers to add into the mesh (from fanout) if they are performing well per `choke_heartbeat_interval`. | 1 |
-| `choke_duplicates_threshold` | The minimum number of duplicates as a percentage of received messages a peer must send before being eligible of being `CHOKE`'d. | 60 |
+| `choke_heartbeat_interval` | The number of heartbeats before assessing and applying `CHOKE`/`UNCHOKE` control messages and adding peers to the mesh. | 20 |
 | `choke_churn` | The maximum number of peers that can be `CHOKE`'d or `UNCHOKE`'d in any `choke_heartbeat_interval`. | 2 |
-|` unchoke_threshold` | Determines how aggressively we unchoke peers. The percentage of messages that we receive in the `choke_heartbeat_interval` that were received by gossip from a choked peer. | 50 |
-| `fanout_addition_threshold` | How aggressively we add peers from the fanout into the mesh. The percentage of messages that we receive in the `choke_heartbeat_interval` that were received from a fanout peer. | 10 |
+|` unchoke_churn` | Determines how aggressively we unchoke peers. The number of peers per `choke_heartbeat_interval` that can be unchoked on an individual mesh. | 2 |
+| `mesh_addition_churn` | How aggressively we add peers from into the mesh. The number of peers per `choke_heartbeat_interval` that can be added to an individual mesh. | 1 |
+
+## Implementation Notes
+
+The actual strategy for choking/unchoking peers is left to each implementation
+and potentially application user. Although the strategies for choking/unchoking
+can be generic, a few useful strategies are listed in the appendix for implementers.
+
+## The CHOKE Message
+
+Every `choke_heartbeat_interval` the router applies its choking strategy to a
+set of collected metrics of recent messages, in order to decide if any of its
+mesh peers should be choked. The router should send no more
+than `choke_churn` `CHOKE` messages to peers per mesh topic. The router should
+also ensure that `D_non_choke` peers remain unchoked in each mesh topic.
+
+Upon receiving a `CHOKE` message, the router MUST no longer forward messages to
+the peer that sent the `CHOKE` message, while it is still in the mesh. Instead
+it MUST always send an IHAVE message (provided it does not hit the IHAVE
+message limit) in the next gossipsub heartbeat to the peer.
```

Comment:
Maybe we could be more aggressive, and send IHAVE (or a similar message) instantly?
This way, the worst delay induced by episub would be 1 RTT, instead of 700ms + 1 RTT (in the case of eth2)

It would also be easier to judge if a chocked peer is faster than unchocked ones

Reply:
Thinking about it, we would even let the "topic score" handle chocking / unchoking:
- When you are fast, you will get "first deliveries" point, and have a big score
- So we choke peers with the lowest scores for this topic (n peers or peers with less than n points)
- If one start to catch up, he will get "first deliveries" points thanks to the IWANT replies
- Up until the point where he isn't amongst the slowest one, and gets unchocked

Reply:
I would rather not conflate scoring with this; scoring is a defense mechanism, while this is a bandwidth optimization mechanism.

Reply:
Also, I've tried to leave the scoring out, because its optional. I'd like episub to work for those users who have opted out of scoring.

In regards to instantly sending the IHAVE, I'm not sure about the best approach. 

I guess benefits of doing it in the heartbeat are:
- We can group messages into a single IHAVE, rather than sending an IHAVE per message
- As you mention, it hads a heartbeat interval plus 1RTT, making decisions about whether to unchoke peers more explicit because they are much faster.  But the downside is we miss peers that are sending messages faster than the mesh but within this timeframe.

Happy to go with either path here. Maybe we try simulate both and see how it goes?

Reply:
That bandwidth is quite easy to compute:

Here is a RPC message with an IHAVE:
`1a0e0a0c0a04626262621204deadbeef`
"deadbeef" is the message id, "26262626" is the topic
So we have 8 bytes of useful data, 8 bytes of headers & co

Even if we add the muxer headers etc, it's still <32 bytes overhead per message, which depending on the use case, might be a lot, or nothing :)

For instance, for eth2 block topic, that's nothing (at most 32 bytes every 12 seconds)
For an attestation topic, chocked peers will send:
- Currently (not chocked): attestation (233 bytes) + topic (~43 bytes) + headers (8) = 284 bytes
- Instant IHAVE: message id (20 bytes) + topic (~43 bytes) + headers (8) = 71 bytes = 75% reduction
- Batched IHAVE: message id (20 bytes) = 20 bytes = 93% reduction
(not counting topic & headers in the batched ihave, as they are enough attestation in one heartbeat that it doesn't matter) 

So we loose 18% reduction on attestations, an about 0% on blocks
(sorry, I'm being eth2 centric here, but I think it highlights two very different use cases)

And of course, if we receive an IHAVE with an unknown block, we will request it, which will cost more bandwidth

Remains to see if these 18% are worth it :) I would say yes as it limit the latency cost quite drastically